README.rst
.. image:: https://travis-ci.org/ome/omero-cli-duplicate.svg?branch=master
    :target: https://travis-ci.org/ome/omero-cli-duplicate

.. image:: https://badge.fury.io/py/omero-cli-duplicate.svg
    :target: https://badge.fury.io/py/omero-cli-duplicate

omero-cli-duplicate
===================

Plugin for use in the OMERO CLI.

Requirements
============

* OMERO 5.6.0 or newer
* Python 3.6 or newer


Installing from PyPI
====================

This section assumes that an OMERO.py is already installed.

Install the command-line tool using `pip <https://pip.pypa.io/en/stable/>`_:

::

    $ pip install -U omero-cli-duplicate

Release process
---------------

This repository uses `bump2version <https://pypi.org/project/bump2version/>`_ to manage version numbers.
To tag a release run::

    $ bumpversion release

This will remove the ``.dev0`` suffix from the current version, commit, and tag the release.

To switch back to a development version run::

    $ bumpversion --no-tag [major|minor|patch]

specifying ``major``, ``minor`` or ``patch`` depending on whether the development branch will be a `major, minor or patch release <https://semver.org/>`_. This will also add the ``.dev0`` suffix.

Remember to ``git push`` all commits and tags.

License
-------

This project, similar to many Open Microscopy Environment (OME) projects, is
licensed under the terms of the GNU General Public License (GPL) v2 or later.

Copyright
---------

2017-2020, The Open Microscopy Environment


CHANGES.txt
=======
Changes
=======

0.4 - Sep 24, 2020
==================
* Further arguments parsed for passing to OMERO.blitz

0.3 - Jan 13, 2020
==================
* Drop support for Python 2

0.2 - Jan 19, 2018
==================
* Update to match https://github.com/ome/omero-cli-render/pull/1

0.1 - April 28, 2016
====================
* Initial release. Based on https://github.com/openmicroscopy/openmicroscopy/pull/4574


test/integration/clitest/test_duplicate.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

#
# Copyright (C) 2016 University of Dundee & Open Microscopy Environment.
# All rights reserved.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

import omero
from omero.plugins.duplicate import DuplicateControl
from cli import CLITest
import pytest

object_types = ["Image", "Dataset", "Project", "Plate", "Screen"]
model = ["", "I"]


class TestDuplicate(CLITest):

    def setup_method(self, method):
        super(TestDuplicate, self).setup_method(method)
        self.cli.register("duplicate", DuplicateControl, "TEST")
        self.args += ["duplicate"]

    def duplicate(self, capfd):
        self.cli.invoke(self.args, strict=True)
        return capfd.readouterr()[0]

    def get_dataset(self, iid):
        """Retrieve all the parent datasets linked to the image"""

        params = omero.sys.ParametersI()
        params.addId(iid)
        query = ("select d from Dataset as d where exists"
                 "(select l from DatasetImageLink as l where "
                 "l.child.id=:id and l.parent=d.id)")
        return self.query.findAllByQuery(query, params)

    def get_project(self, did):
        """Retrieve all the parent projects linked to the dataaset"""

        params = omero.sys.ParametersI()
        params.addId(did)
        query = ("select p from Project as p where exists"
                 "(select l from ProjectDatasetLink as l where "
                 "l.child.id=:id and l.parent=p.id)")
        return self.query.findAllByQuery(query, params)

    @pytest.mark.parametrize("model", model)
    @pytest.mark.parametrize("object_type", object_types)
    def test_duplicate_single_object(self, object_type, model, capfd):
        name = self.uuid()
        oid = self.create_object(object_type, name=name)

        # Duplicate the object
        obj_arg = '%s%s:%s' % (object_type, model, oid)
        self.args += [obj_arg]
        out = self.duplicate(capfd)

        # Check output string
        assert obj_arg in out
        p = omero.sys.ParametersI()
        p.addString("name", name)
        query = "select obj from %s obj where obj.name=:name" % object_type
        objs = self.query.findAllByQuery(query, p)

        # Check object has been duplicated
        assert len(objs) == 2
        assert objs[0].id.val != objs[1].id.val

    def test_duplicate_single_object_dry_run(self, capfd):
        name = self.uuid()
        object_type = "Dataset"
        oid = self.create_object(object_type, name=name)

        # Duplicate the object
        obj_arg = '%s:%s' % (object_type, oid)
        self.args += [obj_arg, "--dry-run"]
        out = self.duplicate(capfd)

        # Check output string
        assert obj_arg in out
        p = omero.sys.ParametersI()
        p.addString("name", name)
        query = "select obj from %s obj where obj.name=:name" % object_type
        objs = self.query.findAllByQuery(query, p)

        # Check original object is present and no duplication happened
        assert len(objs) == 1
        assert objs[0].id.val == oid

    def test_duplicate_single_object_report(self, capfd):
        name = self.uuid()
        object_type = "Dataset"
        oid = self.create_object(object_type, name=name)

        # Duplicate the object
        obj_arg = '%s:%s' % (object_type, oid)
        self.args += [obj_arg, "--report"]
        out = self.duplicate(capfd)
        lines_out = out.splitlines()

        p = omero.sys.ParametersI()
        p.addString("name", name)
        query = "select obj from %s obj where obj.name=:name" % object_type
        objs = self.query.findAllByQuery(query, p)

        # Check object has been duplicated
        assert len(objs) == 2
        assert objs[0].id.val != objs[1].id.val
        assert len(lines_out) == 6
        assert '%s:%s' % (object_type, objs[0].id.val) in out
        assert '%s:%s' % (object_type, objs[1].id.val) in out

    def test_duplicate_single_object_dry_run_report(self, capfd):
        name = self.uuid()
        object_type = "Dataset"
        oid = self.create_object(object_type, name=name)

        # Duplicate the object
        obj_arg = '%s:%s' % (object_type, oid)
        self.args += [obj_arg, "--dry-run", "--report"]
        out = self.duplicate(capfd)
        lines_out = out.splitlines()

        p = omero.sys.ParametersI()
        p.addString("name", name)
        query = "select obj from %s obj where obj.name=:name" % object_type
        objs = self.query.findAllByQuery(query, p)

        # Check object has been duplicated
        assert len(objs) == 1
        assert objs[0].id.val == oid
        assert len(lines_out) == 7
        assert '%s:%s' % (object_type, objs[0].id.val) in out
        # Check object has been found in two different places of the output
        assert out.find(obj_arg) != out.rfind(obj_arg)

    def test_basic_hierarchy_duplication(self, capfd):
        namep = self.uuid()
        named = self.uuid()
        namei = self.uuid()
        proj = self.make_project(namep)
        dset = self.make_dataset(named)
        img = self.make_image(namei)

        self.link(proj, dset)
        self.link(dset, img)

        self.args += ['Project:%s' % proj.id.val]
        self.args += ['--report']
        out = self.duplicate(capfd)

        pp = omero.sys.ParametersI()
        pp.addString("name", namep)
        query = "select obj from Project obj where obj.name=:name"
        objsp = self.query.findAllByQuery(query, pp)

        # Check the Project has been duplicated
        assert len(objsp) == 2
        assert objsp[0].id.val != objsp[1].id.val

        pid = max(objsp[0].id.val, objsp[1].id.val)

        # Check the duplicated Project is in the --report output
        assert 'Project:%s' % pid in out

        pd = omero.sys.ParametersI()
        pd.addString("name", named)
        query = "select obj from Dataset obj where obj.name=:name"
        objsd = self.query.findAllByQuery(query, pd)

        # Check the Dataset has been duplicated
        assert len(objsd) == 2
        assert objsd[0].id.val != objsd[1].id.val

        # Find the Projects linked to the duplicated dataset
        did = max(objsd[0].id.val, objsd[1].id.val)
        proj_linked = self.get_project(did)

        # Check the duplicated Dataset is in the --report output
        assert 'Dataset:%s' % did in out

        # Check the duplicated Dataset is linked to just the duplicated Project
        assert len(proj_linked) == 1
        assert proj_linked[0].id.val == pid

        pi = omero.sys.ParametersI()
        pi.addString("name", namei)
        query = "select obj from Image obj where obj.name=:name"
        objsi = self.query.findAllByQuery(query, pi)

        # Check the image has been duplicated
        assert len(objsi) == 2
        assert objsi[0].id.val != objsi[1].id.val

        # Find the Datasets linked to the duplicated image
        iid = max(objsi[0].id.val, objsi[1].id.val)
        dat_linked = self.get_dataset(iid)

        # Check the duplicated Image is in the --report output
        assert 'Image:%s' % iid in out

        # Check the duplicated image is linked to just the duplicated Dataset
        assert len(dat_linked) == 1
        assert dat_linked[0].id.val == did

    def test_skiphead_duplication(self, capfd):
        namep = self.uuid()
        named = self.uuid()

        proj = self.make_project(namep)
        dset = self.make_dataset(named)

        self.link(proj, dset)

        self.args += ['Project/Dataset:%s' % proj.id.val]
        self.args += ['--report']
        out = self.duplicate(capfd)

        pd = omero.sys.ParametersI()
        pd.addString("name", named)
        query = "select obj from Dataset obj where obj.name=:name"
        objsd = self.query.findAllByQuery(query, pd)

        # Check the Dataset has been duplicated
        assert len(objsd) == 2
        assert objsd[0].id.val != objsd[1].id.val

        did = max(objsd[0].id.val, objsd[1].id.val)

        # Check the duplicated Dataset is in the --report output
        assert 'Dataset:%s' % did in out

        # Find the Projects linked to the duplicated dataset
        proj_linked = self.get_project(did)

        # Check the Dataset is not linked to any Project
        assert len(proj_linked) == 0

        pp = omero.sys.ParametersI()
        pp.addString("name", namep)
        query = "select obj from Project obj where obj.name=:name"
        objsp = self.query.findAllByQuery(query, pp)

        # Check the Project has not been duplicated
        assert len(objsp) == 1


src/omero/plugins/duplicate.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

#
# Copyright (C) 2016-2020 University of Dundee & Open Microscopy Environment.
# All rights reserved.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

"""
   Plugin for duplicating object graphs

"""

import sys

from omero.cli import CLI, GraphControl

HELP = ("""Duplicate graphs of OMERO data based on the ID of the top-node.

By default, a whole subtree of OMERO model objects is duplicated. One
may opt to have duplicate objects reference original parts of the
subtree instead of also duplicating those, see the "--reference" option
below. More strongly, one may "--ignore" given kinds of model objects
such that they are not included among the duplicate in any way; if using
this option then read on to the warning below about ignoring a linked-to
class.

Examples:

    # Duplicate a dataset
    omero duplicate Dataset:50
    # Do the same reporting all the new duplicate objects
    omero duplicate Dataset:50 --report

    # Do a dry run of a duplicate reporting the outcome
    # if the duplicate had been run
    omero duplicate Dataset:53 --dry-run
    # Do a dry run of a duplicate, reporting all the objects
    # that would have been duplicated
    omero duplicate Dataset:53 --dry-run --report

    # Duplicate a project with its datasets but not their images
    omero duplicate Project:15 --ignore DatasetImageLink
    # Duplicate a project with the original images linked from its datasets
    omero duplicate Project:15 --reference Image
"""
        "    # Duplicate a project, linking to the original annotations "
        "except for duplicating the comments and ratings\n"
        "    omero duplicate Project:15 --reference Annotation "
        "--duplicate CommentAnnotation,LongAnnotation\n"
        """
Group permissions can prevent a duplicated link from referencing another
user's Image or Annotation, causing a duplication error. However, using
"--ignore" instead of "--reference" for a linked-to class does not
suffice, one must ignore the link itself. For instance, ignore
ImageAnnotationLink or IAnnotationLink rather than the target
Annotation. This is not an issue for classes such as Roi which can be
ignored directly because they have no separate link class.
""")


class DuplicateControl(GraphControl):

    def cmd_type(self):
        import omero
        import omero.all
        return omero.cmd.Duplicate

    def _pre_objects(self, parser):
        parser.add_argument(
            "--duplicate",
            help="Specify kinds of object to duplicate",
            metavar="CLASS",
            nargs="+", type=lambda s: s.split(","), action="append")
        parser.add_argument(
            "--reference",
            help=("Specify kinds of object to "
                  "link to instead of duplicate"),
            metavar="CLASS",
            nargs="+", type=lambda s: s.split(","), action="append")
        parser.add_argument(
            "--ignore",
            help=("Specify kinds of object to "
                  "ignore, neither linking to nor duplicating"),
            metavar="CLASS",
            nargs="+", type=lambda s: s.split(","), action="append")

    def _process_request(self, req, args, client):
        import omero.cmd
        if isinstance(req, omero.cmd.DoAll):
            requests = req.requests
        else:
            requests = [req]
        for request in requests:
            from itertools import chain
            if isinstance(request, omero.cmd.SkipHead):
                request = request.request
            if args.duplicate:
                request.typesToDuplicate = list(chain(*chain(*args.duplicate)))
            if args.reference:
                request.typesToReference = list(chain(*chain(*args.reference)))
            if args.ignore:
                request.typesToIgnore = list(chain(*chain(*args.ignore)))

        super(DuplicateControl, self)._process_request(req, args, client)

    def print_detailed_report(self, req, rsp, status):
        import omero
        if isinstance(rsp, omero.cmd.DoAllRsp):
            for response in rsp.responses:
                if isinstance(response, omero.cmd.DuplicateResponse):
                    self.print_duplicate_response(response)
        elif isinstance(rsp, omero.cmd.DuplicateResponse):
            self.print_duplicate_response(rsp)

    def print_duplicate_response(self, rsp):
        if rsp.duplicates:
            self.ctx.out("Duplicates")
            obj_ids = self._get_object_ids(rsp.duplicates)
            for k in obj_ids:
                self.ctx.out("  %s:%s" % (k, obj_ids[k]))


try:
    register("duplicate", DuplicateControl, HELP)
except NameError:
    if __name__ == "__main__":
        cli = CLI()
        cli.register("duplicate", DuplicateControl, HELP)
        cli.invoke(sys.argv[1:])
